src/pages/studyView/virtualStudy/VirtualStudy.tsx

```diff
@@ -0,0 +1,248 @@
+        super(props);
+        this.description = getVirtualStudyDescription(props.studyWithSamples, props.selectedSamples, props.filter, this.props.attributeNamesSet, this.props.user);
+    }
+
+    @computed get namePlaceHolder() {
+        return `Selected sample${this.props.selectedSamples.length > 1 ? 's' : ''} (${getCurrentDate()})`
+    }
+
+    @computed get buttonsDisabled() {
+        return _.isEmpty(this.namePlaceHolder) && _.isEmpty(this.name);
+    }
+
+    readonly virtualStudy = remoteData({
+        invoke: async () => {
+            if (this.saving || this.sharing) {
+                let selectedSampleSet = _.groupBy(this.props.selectedSamples, (sample: Sample) => sample.studyId);
+                let studies = _.reduce(selectedSampleSet, (acc: { id: string; samples: string[] }[], samples, studyId) => {
+                    acc.push({
+                        id: studyId,
+                        samples: samples.map(sample => sample.sampleId)
+                    })
+                    return acc;
+                }, []);
+
+                let filters = { patients: {}, samples: {} };
+
+                /* 
+                    TODO: this is to support existing virtual study feature.
+                    but eventually we need to save StudyViewFilter
+                 */
+                let parameters = {
+                    name: this.name || this.namePlaceHolder,
+                    description: this.description,
+                    filters: filters,
+                    origin: this.props.studyWithSamples.map(study => study.studyId),
+                    studies: studies
+                }
+                return await sessionServiceClient.saveVirtualStudy(parameters, this.saving);
+            }
+            return undefined;
+        }
+    }, undefined);
+
+    @computed get virtualStudyUrl() {
+        return getStudySummaryUrl(this.virtualStudy.result ? this.virtualStudy.result.id : '');
+    }
+
+    @autobind
+    private copyLinkRef(el: HTMLAnchorElement | null) {
+        if (el) {
+            new Clipboard(el, {
+                text: () => this.virtualStudyUrl
+            });
+        }
+    }
+
+    @autobind
+    @action onCopyClick() {
+        this.copied = true;
+    }
+
+    @autobind
+    @action
+    private onTooltipVisibleChange(visible: boolean) {
+        this.copied = !visible
+    }
+
+    render() {
+        return (
+            <If condition={this.virtualStudy.isError}>
+                <Then>
+                    <div>
+                        <i className="fa fa-exclamation-triangle" aria-hidden="true" style={{ color: "orange" }} />
+                        <span style={{ marginLeft: "5px" }}>{`Failed to ${this.saving ? 'save' : 'share'} virtual study, please try again later.`}</span>
+                    </div>
+                </Then>
+                <Else>
+                    <If condition={this.virtualStudy.isPending || _.isUndefined(this.virtualStudy.result)}>
```

Comment:
Why do we have an undefined check?

Reply:
default value for `this.virtualStudy.result` is `undefined`